### PR TITLE
Parallel programming helpers

### PR DESCRIPTION
**Pull request by fredrik-johansson**
Turning Flint into Flint-mapreduce!

We could use these helpers in a bunch of places; I'm currently experimenting with them in Arb for pi (binary splitting) and Bernoulli numbers (both parallel_do and binary splitting) and getting nice speedups.

@tthsqe12 could you please check carefully if these functions use the internal threading interface optimally. I copied the binary splitting threading logic from your fmpz_poly_taylor_shift_divconquer.

Things we could add in the future, though I'm keeping it simple in this PR:
* Dynamic scheduling
* A version of parallel do with an interface for each thread to set up its own local state
* Alternative fold strategies (not necessarily binary splitting)
* Not sure why we have both ``thread_pool`` and ``thread_support``; I think we could put all threading helpers in one place.

### CODE REVIEW

**Comment by fredrik-johansson** [comment]
It would be great if the build systems didn't depend on a bunch of filenames being hardcoded in multiple places.

**Comment by tthsqe12** [comment]
Hm, I didn't write that function

**Comment by fredrik-johansson** [comment]
> Hm, I didn't write that function

Indeed, it was Bill who put the thread pool code there.

**Comment by wbhart** [comment]
I'm of course happy to have `thread_pool` and `thread_support` in the one place. They were kept separate because `thread_pool` is supposed to be the *implementation* of the thread pool and `thread_support` is meant to be the user facing functions for threading. But no problems if you want to just combine it all.

I was not sure about the naming convention stuff. I agree that a user facing function should not have `_threaded` attached to its name. The user should get threading automatically if it is selected. However, it may make sense to have a threaded and non-threaded version of the function, each of which is optimised for different user configure options. In this case the internal functions should surely have `_threaded` appended if there is another version of the function for the non-threaded case.

Anyway, I don't recall exactly what we decided on this previously.

I'm ok to merge this once @tthsqe12 has signed off on the parts @fredrik-johansson wanted checked.

Do you want this in the 2.9 release? We are down to 2 tickets which I think we are going to push to the next release, as they are warnings only and seem to us to be (errant) warnings for completely valid code. The only other ticket is to implement division for fmpz_mod_poly and I'm not sure how much I'm going to bother implementing for this release. Not much, perhaps.

**Comment by fredrik-johansson** [comment]
I think it can go in 2.9 if the tests pass everywhere. If you want, I could add a note to the docs saying that the functions are experimental and that the interface may change.

Currently trying this out for a bunch of functions in Arb in https://github.com/fredrik-johansson/arb/tree/threading and it works nicely.

**Comment by wbhart** [comment]
I would say marking it experimental is probably wise. It looks pretty good, but who knows what comes to light after the release.

**Comment by tthsqe12** [comment]
I will have a look at it tomorrow.

**Comment by wbhart** [comment]
Let us know if you want to put an assert in @fredrik-johansson

**Comment by fredrik-johansson** [comment]
Hmm, it's possible to have nt = nw = 1 here.

File: thread_support/thread_support.c [hunk 1]
```diff
@@ -0,0 +1,351 @@
...
+            right_args.init = init;
+            right_args.clear = clear;
+            right_args.args = args;
+            right_args.a = m;
+            right_args.b = b;
+            right_args.basecase_cutoff = basecase_cutoff;
+            right_args.thread_limit = thread_limit;
+            right_args.flags = flags;
+
+            thread_pool_wake(global_thread_pool, threads[0], nt / 2 - 1, _bsplit_worker, &right_args);
```
Comment (tthsqe12) [inline thread_support/thread_support.c:327]: So if the user passes in `thread_limit=1`, we are still requesting one extra worker? And if we get it, waking it up with `max_workers = -1`?

Comment (tthsqe12) [reply thread_support/thread_support.c:327]: For clarification, passing `max_workers=n` here causes the worker threads to have their thread global variable `_flint_num_workers` set to n before they start doing any work, which causes `flint_get_num_threads()` to return `n+1`. So, it looks like `n=-1` could be a problem.
Relevant lines:
https://github.com/wbhart/flint2/blob/trunk/thread_pool/wake.c#L35
https://github.com/wbhart/flint2/blob/trunk/thread_pool/init.c#L29

**Comment by fredrik-johansson** [comment]
It would help if the other thread_support functions like ``flint_request_threads`` were actually documented.

Presumably ``flint_request_threads(&threads, n)`` is meant to return up to ``n - 1`` workers. But if ``nt = 1`` here, we presumably don't want to call it with ``n = 2``.

I just copied the logic from fmpz_poly_taylor_shift_divconquer, so that might need fixing as well.

**Comment by tthsqe12** [comment]
Also, you are spot on with the meaning of `flint_request_threads`.

**Comment by wbhart** [comment]
Is flint_request_threads meant to be used? In the documentation in coding_conventions.txt it is not even mentioned.

The current threading documentation is meant to be experimental until we've ironed the kinks out I think.

**Comment by tthsqe12** [comment]
Well, it might be kinda hard to document the threading interface:
- `flint_set_num_threads` sets the size of global thread pool, and sets the thread global variable `_flint_num_workers`.
- `flint_get_num_threads` does not query the size of the global thread pool, but rather queries the the thread global variable `_flint_num_workers`

**Comment by wbhart** [comment]
Looks like I somehow introduced conflicts. Sorry.

@tthsqe12 What does flint_request_threads do that is different to flint_set_num_threads?

**Comment by fredrik-johansson** [comment]
Anyway, the point of these functions is that they abstract away whatever convoluted things the thread pool API does. Maybe open a separate ticket for the threading documentation?

**Pull request merged**